trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/Boogie2ACSL.java

```diff
@@ -89,6 +95,85 @@ public BacktranslatedExpression translateExpression(
 		return translateExpression(expression, context, false);
 	}
 
+	/**
+	 * Provides a specialized translation for the 'ensures' clause in a procedure contract. Specifically, it adds
+	 * conjuncts of the form 'x == \old(x)' for all global variables that are not modifiable by the procedure, and that
+	 * are off-heap (because we assume conservatively that the procedure may modify the heap).
+	 *
+	 * @param expression
+	 *            The original 'ensures' clause to translate
+	 * @param context
+	 *            The context, i.e., the location of the C function definition
+	 * @param modifiableGlobals
+	 *            The set of variables listed in the Boogie procedure "modifies" clause
+	 * @return A backtranslated ensures clause with the additional conjuncts
+	 */
+	public BacktranslatedExpression translateEnsuresExpression(
+			final de.uni_freiburg.informatik.ultimate.boogie.ast.Expression expression, final ILocation context,
+			final Set<de.uni_freiburg.informatik.ultimate.boogie.ast.IdentifierExpression> modifiableGlobals) {
+		final var intType = new CPrimitive(CPrimitives.INT);
+		final var boolRange = BigInterval.booleanRange();
+
+		if (expression == null) {
+			return computeOldVarEqualities(modifiableGlobals)
+					.map(e -> new BacktranslatedExpression(e, intType, boolRange)).orElse(null);
+		}
+
+		final var mainExpression = translateExpression(expression, context);
+		final var oldVarEqualities = computeOldVarEqualities(modifiableGlobals);
+		if (oldVarEqualities.isPresent()) {
+			return new BacktranslatedExpression(
+					new BinaryExpression(Operator.LOGICAND, mainExpression.getExpression(), oldVarEqualities.get()),
```

Comment:
Check first if `mainExpression` is `null`, i.e., could not be backtranslated.

Reply:
:+1: good catch, done in efe74121b